how-does-fix-with-copilot-work.md

```diff
@@ -0,0 +1,136 @@
+
+In contrast, if you tell Copilot "*This branch can never execute.*", it'll probably find a lot of training that suggests that dead code should be deleted. But it doesn't consider intent or context. Sure, if you're writing an optimiser, dead code elimination is part of the job. But if you end up with a duplicate in an editor, it's more likely that you were copying a long series of cases and forgot to update one of them. (That's why I like the `no-dupe-else-if` rule so much. It's **always** paying attention.)
+
+![Invoking AI](images/ai-no-dupe-if-3.png)
+
+So, specifically for `no-dupe-else-if`, I nudge Copilot to update the copied code. I add "Fix the duplicate condition to be different from the first." to the hidden prompt. Not only does Copilot fix the duplicate case this way, it even gives my nudge as the reason it did so.
+
+![Invoking AI](images/ai-no-dupe-if-2.png)
+
+Sometimes error messages don't actively mislead so much as fail to imply a solution. When you call a function with bad argument types, Typescript says "*Argument of type 'boolean' is not assignable to parameter of type 'number'.*" but doesn't tell you how to fix it. That's on purpose: Typescript has no way to know whether the call or the function itself is wrong.
+
+Copilot might be able to guess, but *Fix With Copilot* can't even see `connect`'s code because it's imported. There's no way for anybody to figure out that the arguments are swapped. So, with just the error message to go on, Copilot makes a very local change: it reasons that `0` is the closest `number` equivalent of `false`. That technically fixes the error but another immediately pops up, because it's not a correct fix.
+
+![Invoking AI](images/ai-bad-call-3.png)
+
+When I look up and include the code for `connect` in the hidden prompt, Copilot can figure out that the arguments are swapped.
+
+![Invoking AI](images/ai-bad-call-2.png)
+
+## But really, how does Fix With Copilot work?
+
+The first thing *Fix With Copilot* does is register as a fix provider for every kind of file. Every single one. Then, for each document you have open, it requests errors and adds *Fix With Copilot* to the list of available fixes for each.
+<!-- see inlineChatCommand.ts:registerInlineChatCommands
+and inlineChatCodeActions.ts:QuickFixesProvider -->
+
+When you click on *Fix With Copilot*, it invokes Copilot's inline chat with a hidden prompt constructed from JSX components. JSX&mdash;but not React. Copilot has its own renderer that generates plain text from JSX components, so *Fix With Copilot* is structured as a React-like hierarchy of small components. In addition to looking familiar, one clever property on all these components is `priority`. The renderer can drop low-priority components if the prompt gets too big. Here's an example of what the code looks like. `diagnostics` is the list of errors to be fixed:
```

Comment:
There's a bit of a jump here - you might want to talk about how we could just stitch together raw text by writing our prompts in strings, JSX gives little bit of visual ergonomics and abstraction over priority.

On that note, priority doesn't come up again after you mention it - what in the below example is deprioritized?

Reply:
Hmph. Priorities are only specified for major blocks, like the whole block of diagnostic generation that I'm talking about. Maybe I'll omit that and instead explain the code a little bit.